test/unistd/links.c

```diff
@@ -59,56 +100,131 @@ int main() {
+void test_overwriting_symlink()
+{
+  int rtn = symlink("new-nonexistent-path", "link");
+  assert(rtn == -1);
   assert(errno == EEXIST);
   errno = 0;
+}
 
-  printf("\nsymlink/normal\n");
-  ret = symlink("new-nonexistent-path", "folder/link");
-  assert(ret == 0);
+void test_creating_symlink()
+{
+  int rtn = symlink("new-nonexistent-path", "directory/link");
+  assert(rtn == 0);
   assert(errno == 0);
   errno = 0;
 
-  printf("\nreadlink(created link)\n");
-  ret = readlink("folder/link", buffer, 256);
+  char buffer[256] = {0};
+  rtn = readlink("directory/link", buffer, 256);
   assert(errno == 0);
 #if !defined(__EMSCRIPTEN__) || defined(WASMFS)
   assert(strcmp(buffer, "new-nonexistent-path") == 0);
 #else
-  assert(strcmp(buffer, "/working/folder/new-nonexistent-path") == 0);
+  assert(strcmp(buffer, "/working/directory/new-nonexistent-path") == 0);
 #endif
-  assert(strlen(buffer) == ret);
+  assert(strlen(buffer) == rtn);
   errno = 0;
+}
 
+void test_reading_shortened_symlink()
+{
+  char buffer[256] = {0};
+  readlink("directory/link", buffer, 256);
   buffer[0] = buffer[1] = buffer[2] = buffer[3] = buffer[4] = buffer[5] = '*';
-  printf("\nreadlink(short buffer)\n");
-  ret = readlink("link", buffer, 4);
+  int rtn = readlink("link", buffer, 4);
   assert(errno == 0);
-  assert(ret == 4);
+  assert(rtn == 4);
 #if !defined(__EMSCRIPTEN__) || defined(WASMFS)
   assert(strcmp(buffer, "../t**nexistent-path") == 0);
 #else
-  assert(strcmp(buffer, "/the**ng/folder/new-nonexistent-path") == 0);
+  assert(strcmp(buffer, "/the**ng/directory/new-nonexistent-path") == 0);
 #endif
   errno = 0;
+}
 
+void test_noticing_loop_in_symlink()
+{
   // FS.lookupPath should notice the symlink loop and return ELOOP, not go into
   // an infinite recurse.
   //
   // This test doesn't work in wasmfs -- probably because access sees the
   // symlink and returns true without bothering to chase the symlink
-  symlink("./linkX/inside", "./linkX");
-  ret = access("linkX", F_OK);
-  assert(ret == -1);
+  symlink("./loop-link/inside", "./loop-link");
+  int rtn = access("loop-link", F_OK);
+  assert(rtn == -1);
   assert(errno == ELOOP);
   errno = 0;
+}
+
+
+void test_relative_path_symlinks()
+{
+  char* parents[] = {
+    "/working/directory/",
+    "/working/directory/subdirectory/",
+    "/working/"
+  };
 
+  char* links[] = {
+    "relative",
+    "subrelative",
+    "subdirectoryrelative",
+  };
+
+  for (int i = 0; i < sizeof links / sizeof links[0]; i++) {
+    int rtn = chdir(parents[i]);
+    assert(rtn == 0);
+    char symlink[256] = {0};
+    strcat(strcpy(symlink, parents[i]), links[i]);
+    printf("symlink: '%s'\n", symlink);
+    char buf[256] = {0};
+    rtn = readlink(links[i], buf, 256);
+    FILE *fd = fopen(buf, "r");
+    assert(fd);
+    char buffer[13] = {0};
+    rtn = fread(buffer, 1, 13, fd);
+    assert(rtn <= 13);
+    printf("buffer: '%s'\n\n", buffer);
+    fclose(fd);
+  }
+}
+
+void test_absolute_path_symlinks()
```

Comment:
Can you put the opening curly braces on the same line as the function decl?

Reply:
Sorry, on it.